contents/javascript/Javascript-promises.md

```diff
@@ -0,0 +1,129 @@
+
+In Javascript, there are times where an operation is executed asynchronously and its result may not be obtained immediately, unlike synchronous functions with a return value. It can be especially confusing to trace the code as it may not run in the sequence that you would normally expect.
+
+We'll use `setTimeout()` as an example of this behaviour. This is an asynchronous function takes in 2 parameters, a callback and a time (in milliseconds) to wait before the callback is executed. A callback is a function that is passed as a parameter to another function (call this A), and it will be executed after A finishes. 
+
+The following code snippet logs `Hello World` after 1 second has elapsed. The function that encapsulates the `console.log()` is the callback function.
+
+```javascript
+
+setTimeout(function (){
+	console.log("Hello World");
+}, 1000);
+
+```
+
+Javascript is at its core, a synchronous and single-threaded language but it is able to provide the illusion of asynchronicity through a construct known as the Event Loop. This is a loop that allows asynchronous functions to be queued for execution later on. Further information about the Event Loop can be found at [Mozilla's Event Loop documentation](https://developer.mozilla.org/en-US/docs/Web/JavaScript/EventLoop).
+
+Another pertinent example of asynchronicity is HTTP requests. When making a HTTP request, the information from the server is not instantly returned as there is a round trip time that involves the time taken for the request to be sent, and for the response from the server to be sent back to the client. In this time, Javascript would not necessarily wait for the response to be received, but instead execute other functions first, even if those functions depend on the result of that HTTP response.
+
+In the code snippet below, `console.log(price)` would return undefined as the asynchronous `retrievePrice()` function has not finished executing.
+
+```javascript
+var price;
+price = retrievePrice("https://www.amazon.com/gp/product/B009S331VU/");
+console.log(price);
+
+```
+
+Actual Execution Flow:
+![Image](asyncRequestSequence.png)
+
+Since `retrivePrice()` is a non-blocking function and it's only handled after the `console.log()` has executed, there must be a way to ensure that these functions are executed in the sequence that's desired.
+
+This is where Promises come in handy.
+
+
+## Introduction
+Promises in Javscript behave the same way as Promises do in real life. It is a proxy for a value that is not known at the time of Promise creation. It has 3 states: pending, fulfilled and rejected. Promises provide the power to control the execution flow of your program by allowing the user to wait for a promise to be resolved before carrying on with the rest of the program execution.
+
+## How to use Promises
```

Comment:
May be reframe this as 'How Promises are used' and give some nice examples of how Promises are used? The current title suggests that the section teaches how to use Promises. Normally our learning resources don't try to teach topics. Rather, we want to motivate the reader to learn the topic and point to other resources for the actual learning of the topic.